### PR TITLE
1592006: Implement events for Python bindings

### PR DESCRIPTION
**Pull request by mdboom**

### CODE REVIEW

File: glean-core/python/glean/_loader.py [hunk 1]
```diff
@@ -87,7 +92,15 @@ def _get_metric_object(metric: glean_parser.metrics.Metric) -> Any:
 
     glean_metric.__doc__ = metric.description
 
-    return glean_metric
+    yield name, glean_metric
+
+    # Events also need to define an enumeration
+    if metric.type == "event":
```
Comment (Dexterp37) [inline glean-core/python/glean/_loader.py]: I wish we could just use glean_parser for these kind of things :(

File: glean-core/python/glean/util.py [hunk 2]
```diff
@@ -22,3 +23,10 @@ def get_locale_tag() -> str:
     value = value.replace("_", "-")
 
     return value
+
+
+def time_ms() -> int:
```
Comment (Dexterp37) [inline glean-core/python/glean/util.py]: I bet there's no concept of `uint`, right?

File: glean-core/python/glean/_ffi.py [hunk 3]
```diff
@@ -98,6 +98,13 @@ def ffi_encode_vec_string(strings: List[str]) -> Any:
     return result
 
 
+def ffi_encode_vec_int32(values: List[str]) -> Any:
```
Comment (badboy) [inline glean-core/python/glean/_ffi.py]: `List[str]`? Shouldn't this be `List[int]`?

File: glean-core/python/glean/metrics/event.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                where additional richer context is needed. The maximum length
+                for values is 100.
+        """
+        if self._disabled:
+            return
+
+        timestamp = util.time_ms()
+
+        @Dispatcher.launch
+        def record():
```
Comment (badboy) [inline glean-core/python/glean/metrics/event.py]: I hope this just shadows the above function definition and thus won't cause any problems?

File: glean-core/python/glean/_ffi.py (hunk 3 above)
Comment (mdboom) [reply glean-core/python/glean/_ffi.py]: Good catch.

File: glean-core/python/glean/_loader.py (hunk 1 above)
Comment (mdboom) [reply glean-core/python/glean/_loader.py]: Not sure I follow.  We *are* using glean_parser here.  This file is essentially what is required for any of the language generators (just generating objects, not code).  The parsing etc. still uses glean_parser underneath.

File: glean-core/python/glean/metrics/event.py (hunk 4 above)
Comment (mdboom) [reply glean-core/python/glean/metrics/event.py]: It's a nested function, so not accessible from the outside.  We could really call it whatever we want -- it's never even used by name -- it's just a way to put a function on the dispatcher.  If Python had anonymous functions, I'd use that...

File: glean-core/python/glean/util.py (hunk 2 above)
Comment (mdboom) [reply glean-core/python/glean/util.py]: That's correct: Python 3 only has a single `int` type.

File: glean-core/python/glean/_loader.py (hunk 1 above)
Comment (Dexterp37) [reply glean-core/python/glean/_loader.py]: Yup, sorry, let me clarify: I wish this was as simple as just creating a `jinja` template in glean_parser :)

File: glean-core/python/glean/metrics/event.py (hunk 4 above)
Comment (badboy) [reply glean-core/python/glean/metrics/event.py]: 👍

**Pull request merged**